src/execution_plan/ops/op_node_hash_join.c

```diff
@@ -0,0 +1,244 @@
+    op->number_of_intersections--;
+
+    return cr;
+}
+
+static uint64_t _binarySearch(const OpNodeHashJoin *op, NodeID id) {
+    Record *array = op->cached_records;
+    uint32_t recordCount = array_len(array);
+    uint32_t pos;
+    uint32_t left = 0;
+    uint32_t right = recordCount;
+    while(left < right) {
+        pos = (right + left) / 2;
+        Node *n = Record_GetNode(array[pos], op->intersect_node_idx);
+        if(ENTITY_GET_ID(n) < id) left = pos + 1;
+        else right = pos;
+    }
+    return left;
+}
+
+/* Look up first cached record CR position
+ * where CR node ID = id.
+ * Returns NO_INTERSECTION if no such CR is found. */
+int _set_intersection_idx(OpNodeHashJoin *op, NodeID id) {
+    op->intersect_idx = NO_INTERSECTION;
+    op->number_of_intersections = 0;
+    int idx = _binarySearch(op, id);
+
+    // Make sure node was found.
+    Record r = op->cached_records[idx];
+    Node *n = Record_GetNode(r, op->intersect_node_idx);
+
+    // Node was not found.
+    if(ENTITY_GET_ID(n) != id) return NO_INTERSECTION;
+
+    /* Node was found, make sure op->intersect_idx
+     * points to the first record holding node, 
+     * and update number_of_intersections to count how many
+     * records share the same node ID. */
+    while(idx >= 0) {
+        r = op->cached_records[idx];
+        Node *n = Record_GetNode(r, op->intersect_node_idx);
+        if(ENTITY_GET_ID(n) != id) break;
+        idx--;
+    }
+
+    op->intersect_idx = idx;
+    if(idx > 0) op->intersect_idx++;
+    idx = op->intersect_idx;
+
+    r = op->cached_records[idx];
+    n = Record_GetNode(r, op->intersect_node_idx);
+    uint record_count = array_len(op->cached_records);
+
+    // Count how many records share the same node.
+    while(idx < record_count) {
+        r = op->cached_records[idx];
+        n = Record_GetNode(r, op->intersect_node_idx);
+        if(ENTITY_GET_ID(n) != id) break;
+        op->number_of_intersections++;
+    }
+
+    return op->intersect_idx;
+}
+
+/* Caches all records coming from left branch. */
+void _cache_records(OpNodeHashJoin *op) {
+    assert(op->cached_records == NULL);
+
+    OpBase *left_child = op->op.children[0];
+    op->cached_records = array_new(Record, 32);
+
+    Record l = NULL;
+    while((l = left_child->consume(left_child)))
+        op->cached_records = array_append(op->cached_records, l);
+}
+
+void _sort_cached_records(OpNodeHashJoin *op) {
+    uint intersect_node_idx = op->intersect_node_idx;
+    QSORT(Record, op->cached_records,
+          array_len(op->cached_records), RECORD_SORT_ON_ENTRY);
+}
+
+static int NodeHashJoinToString(const OpBase *ctx, char *buff, uint buff_len) {
+    const OpNodeHashJoin *op = (const OpNodeHashJoin*)ctx;
+
+    int offset = 0;    
+    offset += snprintf(buff + offset, buff_len-offset, "%s | ", op->op.name);
+    offset += Node_ToString(op->intersect_node, buff + offset, buff_len - offset);
+    return offset;
+}
+
+/* Creates a new NodeHashJoin operation */
```

Comment:
This is a fairly obvious point, but we're not performing a hash join here! Do we need this particular name for TCK coverage?